testsuite/tests/flambda-unboxing/rec_unboxing.ml

```diff
@@ -0,0 +1,240 @@
+     value of [aux] should be allocated, and not any of the floats inside of
+     the loop; thus by having a high loop count, we can check whether the float
+     in the loop was unboxed by verifying that the number of words allocated is
+     less than a small constant. *)
+  let[@inline] aux n y =
+    let r = ref y in
+    for i = 0 to n - 1 do
+      r := !r +. 1.
+    done;
+    !r
+  in
+  check_not_many_allocs 5. "float-hard-1" aux 1_000 0. 1_000.;
+  check_not_many_allocs 5. "float-hard-2" aux 10_000 0. 10_000.;
+  (* Check tuple.
+     Not currently unboxed. *)
+  let aux n ((_x, _y) as t) res =
+    let r = ref t in
+    for i = 0 to n do
+      let x, y = !r in
+      r := y, x
+    done;
+    let x, y = !r in
+    x - y = res
+  in
+  check_no_alloc "tuple-loop-1" aux 0 (1,2) 1;
+  check_no_alloc "tuple-loop-2" aux 1 (1,2) ~-1;
+  (* Check nested tuples.
+     Not currently unboxed. *)
+  let aux n (((_x, _y), _z) as t) res =
+    let r = ref t in
+    for i = 0 to n do
+      let ((x, y), z) = !r in
+      r := (y, z), x
+    done;
+    let (x, y), z = !r in
+    x + (y - z) = res
+  in
+  check_no_alloc "tuple-loop-3" aux 0 ((1,2),3) 4;
+  check_no_alloc "tuple-loop-4" aux 1 ((1,2),3) 2;
+  (* Check tuple+floats.
+     Not unboxed yet. *)
+  let aux n ((_x, _y) as t) res =
+    let r = ref t in
+    for i = 0 to n do
+      let x, y = !r in
+      r := x +. y, x -. y
+    done;
+    let x, y = !r in
+    x +. y = res
+  in
+  check_no_alloc "tuple-float-1" aux 0 (1.,2.) 2.;
+  check_no_alloc "tuple-float-2" aux 1 (1.,2.) 6.;
+  (* Check blocks and floats.
+     Not unboxed yet. *)
+  let aux n (c1, c2) res =
+    let r = ref c1 in
+    for i = 0 to n do
+      r := (Complex.add[@inlined]) !r c2;
+    done;
+    let f =
+      if n mod 2 = 0
+      then (Complex.arg[@inlined]) !r
+      else (Complex.norm[@inlined]) !r
+    in
+    f = res
+  in
+  check_no_alloc "complex-loop-1" aux 0 (Complex.zero, Complex.one) 0.;
+  check_no_alloc "complex-loop-2" aux 1 (Complex.zero, Complex.one) 2.;
+  (* Check variants.
+     Not unboxed yet. *)
+  let aux n x res =
+    let r = ref None in
+    for i = 0 to n do
+      match !r with
+      | None -> r := Some x
+      | Some _ -> r := None
+    done;
+    let r =
+      match !r with
+      | None -> 13
+      | Some x -> x
+    in
+    r = res
+  in
+  check_no_alloc "variant-loop-1" aux 0 0 0;
+  check_no_alloc "variant-loop-2" aux 1 0 13;
+  (* Check more complicated variant.
+     Not unboxed yet. *)
+  let aux n (a, b) res =
+    let r = ref T.A in
+    for i = 0 to n do
+      match (!r : T.t) with
+      | A -> r := B
+      | B -> r := C a
+      | C _ -> r := D (a, b)
+      | D _ -> r := A
+    done;
+    let e =
+      match (!r : T.t) with
+      | C a -> a
+      | D (a, b) -> a + b
+      | A -> 42
+      | B -> 50
+    in
+    e = res
+  in
+  check_no_alloc "variant-loop-3" aux 0 (1,2) 50;
+  check_no_alloc "variant-loop-4" aux 1 (1,2) 1;
```

Comment:
Maybe rename the result as `"variant-loop-4-expecting-to-fail-for-now"` so that it is easier to check when looking at the output file

Reply:
And this applies for the other cases above too.

Reply:
done